src/Build/Logging/BinaryLogger/BinaryLogger.cs

```diff
@@ -0,0 +1,136 @@
+        /// </summary>
+        public string Parameters { get; set; }
+
+        /// <summary>
+        /// Initializes the logger by subscribing to events of IEventSource
+        /// </summary>
+        public void Initialize(IEventSource eventSource)
+        {
+            Environment.SetEnvironmentVariable("MSBUILDTARGETOUTPUTLOGGING", "true");
+
+            ProcessParameters();
+
+            try
+            {
+                stream = new FileStream(FilePath, FileMode.Create);
+            }
+            catch (Exception e)
+            {
+                string errorCode;
+                string helpKeyword;
+                string message = ResourceUtilities.FormatResourceString(out errorCode, out helpKeyword, "InvalidFileLoggerFile", FilePath, e.Message);
+                throw new LoggerException(message, e, errorCode, helpKeyword);
+            }
+
+            stream = new GZipStream(stream, CompressionLevel.Optimal);
+            binaryWriter = new BinaryWriter(stream);
+            eventArgsWriter = new BuildEventArgsWriter(binaryWriter);
+
+            binaryWriter.Write((byte)FileFormatVersion);
+
+            eventSource.AnyEventRaised += EventSource_AnyEventRaised;
+        }
+
+        /// <summary>
+        /// Closes the underlying file stream.
+        /// </summary>
+        public void Shutdown()
+        {
+            if (stream != null)
+            {
+                // It's hard to determine whether we're at the end of decoding GZipStream
+                // so add an explicit 0 at the end to signify end of file
+                stream.WriteByte(0);
+                stream.Flush();
+                stream.Dispose();
+                stream = null;
+            }
+        }
+
+        private void EventSource_AnyEventRaised(object sender, BuildEventArgs e)
+        {
+            Write(e);
+        }
+
+        private void Write(BuildEventArgs e)
+        {
+            if (stream != null)
+            {
+                // TODO: think about queuing to avoid contention
+                lock (eventArgsWriter)
+                {
+                    eventArgsWriter.Write(e);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Processes the parameters given to the logger from MSBuild.
+        /// </summary>
+        /// <exception cref="LoggerException">
+        /// </exception>
+        private void ProcessParameters()
+        {
+            const string invalidParamSpecificationMessage = @"Need to specify a valid log file name, such as msbuild.binlog";
+
+            if (Parameters == null)
+            {
+                throw new LoggerException(invalidParamSpecificationMessage);
+            }
+
+            string[] parameters = Parameters.Split(';');
+
+            if (parameters.Length != 1)
+            {
+                throw new LoggerException(invalidParamSpecificationMessage);
+            }
+
+            FilePath = parameters[0].TrimStart('"').TrimEnd('"');
```

Comment:
Technically it's not even needed since the arg parser does it already. This is just for the case someone instantiates a logger outside of MSBuild.